### PR TITLE
Gradle magically creates tasks on findByName

### LINKED ISSUE #156: tasks.findByName magically *creates* tasks, preventing future mutation (!!)
after hours of debugging the entirely unhelpful message:
```
Caused by: java.lang.IllegalStateException: Cannot add Mutate rule 
'org.gradle.language.base.plugins.ComponentModelBasePlugin$Rules#closeSourcesForBinaries(org.gradle.platform.base.BinaryContainer, org.gradle.language.base.ProjectSourceSet)'
for model element 'binaries' when element is in state GraphClosed.
```

the issue is the use of findByName('assemble').  This cause Gradle to *create* the assembly task, preventing modification of say the binaries later.

Pulpit:
This is why I hate things that have too much magic.  Only a line by line binary search caused to me find this, since the stack trace is thrown much later when the project's tasks have to be resolved, and nowhere near either the mutation or the findByName.  SHEESH!

@brunobowden I'm fixing this to allow native dependencies, hopefully you can rollback your workaround for plugin syntax once I merge this in.

### PR DESCRIPTION
**Pull request by advayDev1**
Gradle magically creates tasks on findByName
which prevents you from modifying their inputs
later on.

Instead use tasks.all which seems not to have this issue.

Fixes #156

### CODE REVIEW

**Comment by brunobowden** [comment]
So you can reorder the 'j2objc' plugin to be applied before the 'java' plugin and it still works?

Looks good assuming the above is true. I guess this'll take a new version spin to test properly with the `plugin {}` syntax.

**Comment by advayDev1** [comment]
I haven't tested that; what I did test was that a series of other issues regarding StateClosed were fixed.  (I'm trying to do #51 and it needs modifying the native binary config.)

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 1]
```diff
@@ -186,14 +186,17 @@ class J2objcPlugin implements Plugin<Project> {
     // The before task must already exist.
     private def lateDependsOn(Project proj, String afterTaskName, String beforeTaskName) {
         assert null != proj.tasks.findByName(beforeTaskName)
-        def afterTask = proj.tasks.findByName(afterTaskName)
-        if (afterTask != null) {
-            afterTask.dependsOn beforeTaskName
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: @bigguy suggested to try this...
```
private def lateDependsOn(Project proj, String afterTaskName, String beforeTaskName) {
    assert null != proj.tasks.findByName(beforeTaskName)
    if (afterTask != null) {
        afterTask.dependsOn {
            // evaluated after all tasks are created
            proj.tasks.findByName(beforeTaskName)
        }
    }
}
```

Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: I'm OK with either.  They are equally confusing to me but this new
suggestion puts the depends on upfront so I'll do that.

Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Sorry no that doesn't work.  It is after task which is the one we do not
want to materialize.  That soln requires after task to be found inline.

**Comment by advayDev1** [comment]
due to my response above, is this LG to submit?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 2]
```diff
@@ -186,14 +186,17 @@ class J2objcPlugin implements Plugin<Project> {
...
+        // reasons unknown, the Java plugin creates - right there! - the task; this prevents
+        // later code from modifying binaries, sourceSets, etc.  If you see an error
+        // mentioning 'state GraphClosed' saying you can't mutate some object, see if you are magically
+        // causing Gradle to make the task by using findByName!  Issue #156
+
+        // tasks.all cleanly calls this closure on any existing elements and for all elements
+        // added in the future.
+        proj.tasks.all { task ->
+            if (task.name == afterTaskName) {
+                task.dependsOn beforeTaskName
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: LGTM

Post something on StackOverflow to see if there's a better way of doing this and add a TODO. I'd prefer not to iterate through all the tasks by name, it feels like there should be a better way. If Gradle isn't responsive to the SO question, I can give them a nudge over email.

**Comment by advayDev1** [comment]
will do, merging.

**Comment by confile** [comment]
LGTM

**Pull request merged**